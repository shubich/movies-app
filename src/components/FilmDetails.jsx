import React from 'react';
import { imagePrefix } from '../api/'

export class FilmDetails extends React.Component {
    constructor(props) {
        super(props);
        if (!this.props.fetching) {
            this.props.getFilmDetails();            
            this.props.getSimilarFilms();            
        }
    }

    componentWillReceiveProps(nextProps) {
        let currentId =  this.props.match.params.id;
        let nextId = nextProps.match.params.id;

        if (currentId != nextId) {           
            this.props.getFilmDetails(nextId);
            this.props.getSimilarFilms(nextId);
        }
    }
    
    render() {
        let details = this.props.details;
        
        return (
            details.title ?
            <div>
                <div id='film-details'>
                    <div className='poster'>
                        <img src={imagePrefix+details.poster_path} alt="poster"/>
                    </div>
                    <div className='info'>
                        <h2>
                            <span className='title text-primary'>{details.title}</span>
                            <span className='rating'>{details.vote_average}</span>
                        </h2>
                        <p className='genre text-basic'>{details.genres}</p>
                        <p className='numbers'>
                            <span className='release-year'>{details.release_date}</span>
                            <span className='duration'>{details.runtime}</span>
                        </p>
                        <p className='description text-basic'>{details.overview}</p>
                        {/* <p className='director text-basic'>Director: Quentin Tarantino</p>
                        <p className='cast text-basic'>Cast: [...actors]</p> */}
                    </div>
    
                </div>
            </div>
            :
            null
        )
    }
}