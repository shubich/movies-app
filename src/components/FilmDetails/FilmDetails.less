@import "../../styles/palette";
@import "../../styles/media";

.FilmDetails-Genres {
    color: @basic-color;
}

.FilmDetails-Numbers {
    font-weight: bold;
    span {
        margin-right: 25px;
    }
}

.FilmDetails-Description {
    margin-top: 1.4em;
    font-size: 1.1em;
    color: @basic-color;
    white-space: pre-line;
}

@media @phone {
    .FilmDetails-Description {
        margin-top: 0.9em;
    }
}
