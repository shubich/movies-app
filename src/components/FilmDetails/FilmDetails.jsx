import React from 'react';
import PropTypes from 'prop-types';
import Poster from '../Poster';
import Details from '../Details';

import './FilmDetails.less';

const FilmDetails = (props) => {
  if (!props.details.title) return null;

  return (
    <Details fetching={props.fetching}>
      <Poster
        className="FilmDetails-Poster Details-Poster"
        path={props.details.poster_path}
        film
      />
      <div className="FilmDetails-Info">
        <h2 className="FilmDetails-Header Details-Header">
          <span className="FilmDetails-Title Details-Title">
            {props.details.title}
          </span>
          <span className="FilmDetails-Rating Details-Rating">
            {props.details.vote_average}
          </span>
        </h2>
        <p className="FilmDetails-Genres">
          {props.details.genres.map(genre => genre.name).join(', ')}
        </p>
        <p className="FilmDetails-Numbers">
          {
            props.details.release_date &&
            <span className="FilmDetails-ReleaseYear">
              {
                new Date(props.details.release_date)
                .toLocaleDateString('en-US', {
                  year: 'numeric', month: 'short', day: 'numeric',
                })
              }
            </span>
          }
          {
            props.details.runtime ? (
              <span className="FilmDetails-Duration">
                {props.details.runtime} min
              </span>
            ) : (null)
          }
        </p>
        <p className="FilmDetails-Description">
          {props.details.overview}
        </p>
      </div>
    </Details>
  );
};

export default FilmDetails;

FilmDetails.propTypes = {
  fetching: PropTypes.bool.isRequired,
  details: PropTypes.shape({
    poster_path: PropTypes.string,
    title: PropTypes.string,
    vote_average: PropTypes.number,
    genres: PropTypes.array,
    release_date: PropTypes.string,
    runtime: PropTypes.number,
    overview: PropTypes.string,
  }).isRequired,
};
