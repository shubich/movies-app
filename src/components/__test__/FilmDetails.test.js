import React from 'react';
import { FilmDetails } from '../FilmDetails';
import { shallow } from 'enzyme';

let props;

describe('FilmDetails', () => {
    beforeEach(() => {
        props = {
            details: {
                title: 'Pulp Fiction',
                poster_path: '/img1234_0987',
                release_date: '1999',
                vote_average: '8.3',
                genres: ['Thriller', 'Crime'],
                runtime: '154',
                overview: 'Lorem ipsum dolor sit amet.'
            },
            match: {
                params: {
                    id: '680'
                }
            },
            getFilmDetails: jest.fn,
            getSimilarFilms: jest.fn
        };
    });

    it('renders correctly with details', () => {
        const wrapper = shallow(
            <FilmDetails {...props} />
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('renders correctly without details', () => {
        const wrapper = shallow(
            <FilmDetails {...props} details={[]} />
        );
        expect(wrapper).toMatchSnapshot();
    });
})