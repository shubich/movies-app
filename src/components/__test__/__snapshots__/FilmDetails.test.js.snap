// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilmDetails renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FilmDetails
    details={
      Object {
        "genres": Array [
          "Thriller",
          "Crime",
        ],
        "overview": "Lorem ipsum dolor sit amet.",
        "poster_path": "/img1234_0987",
        "release_date": "1999",
        "runtime": "154",
        "title": "Pulp Fiction",
        "vote_average": "8.3",
      }
    }
    getFilmDetails={[Function]}
    getSimilarFilms={[Function]}
    match={
      Object {
        "params": Object {
          "id": "680",
        },
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "host",
    "props": Object {
      "children": <div
        id="film-details"
      >
        <div
          className="poster"
        >
          <img
            alt="poster"
            src="https://image.tmdb.org/t/p/w640/img1234_0987"
          />
        </div>
        <div
          className="info"
        >
          <h2>
            <span
              className="title text-primary"
            >
              Pulp Fiction
            </span>
            <span
              className="rating"
            >
              8.3
            </span>
          </h2>
          <p
            className="genre text-basic"
          >
            Thriller
            Crime
          </p>
          <p
            className="numbers"
          >
            <span
              className="release-year"
            >
              1999
            </span>
            <span
              className="duration"
            >
              154
            </span>
          </p>
          <p
            className="description text-basic"
          >
            Lorem ipsum dolor sit amet.
          </p>
        </div>
      </div>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": null,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="poster"
          >
            <img
              alt="poster"
              src="https://image.tmdb.org/t/p/w640/img1234_0987"
            />
          </div>,
          <div
            className="info"
          >
            <h2>
              <span
                className="title text-primary"
              >
                Pulp Fiction
              </span>
              <span
                className="rating"
              >
                8.3
              </span>
            </h2>
            <p
              className="genre text-basic"
            >
              Thriller
              Crime
            </p>
            <p
              className="numbers"
            >
              <span
                className="release-year"
              >
                1999
              </span>
              <span
                className="duration"
              >
                154
              </span>
            </p>
            <p
              className="description text-basic"
            >
              Lorem ipsum dolor sit amet.
            </p>
          </div>,
        ],
        "id": "film-details",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": <img
              alt="poster"
              src="https://image.tmdb.org/t/p/w640/img1234_0987"
            />,
            "className": "poster",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": null,
            "nodeType": "host",
            "props": Object {
              "alt": "poster",
              "src": "https://image.tmdb.org/t/p/w640/img1234_0987",
            },
            "ref": null,
            "rendered": null,
            "type": "img",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h2>
                <span
                  className="title text-primary"
                >
                  Pulp Fiction
                </span>
                <span
                  className="rating"
                >
                  8.3
                </span>
              </h2>,
              <p
                className="genre text-basic"
              >
                Thriller
                Crime
              </p>,
              <p
                className="numbers"
              >
                <span
                  className="release-year"
                >
                  1999
                </span>
                <span
                  className="duration"
                >
                  154
                </span>
              </p>,
              <p
                className="description text-basic"
              >
                Lorem ipsum dolor sit amet.
              </p>,
            ],
            "className": "info",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": null,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <span
                    className="title text-primary"
                  >
                    Pulp Fiction
                  </span>,
                  <span
                    className="rating"
                  >
                    8.3
                  </span>,
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": null,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Pulp Fiction",
                    "className": "title text-primary",
                  },
                  "ref": null,
                  "rendered": "Pulp Fiction",
                  "type": "span",
                },
                Object {
                  "instance": null,
                  "key": null,
                  "nodeType": "host",
                  "props": Object {
                    "children": "8.3",
                    "className": "rating",
                  },
                  "ref": null,
                  "rendered": "8.3",
                  "type": "span",
                },
              ],
              "type": "h2",
            },
            Object {
              "instance": null,
              "key": null,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "Thriller",
                  "Crime",
                ],
                "className": "genre text-basic",
              },
              "ref": null,
              "rendered": Array [
                "Thriller",
                "Crime",
              ],
              "type": "p",
            },
            Object {
              "instance": null,
              "key": null,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <span
                    className="release-year"
                  >
                    1999
                  </span>,
                  <span
                    className="duration"
                  >
                    154
                  </span>,
                ],
                "className": "numbers",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": null,
                  "nodeType": "host",
                  "props": Object {
                    "children": "1999",
                    "className": "release-year",
                  },
                  "ref": null,
                  "rendered": "1999",
                  "type": "span",
                },
                Object {
                  "instance": null,
                  "key": null,
                  "nodeType": "host",
                  "props": Object {
                    "children": "154",
                    "className": "duration",
                  },
                  "ref": null,
                  "rendered": "154",
                  "type": "span",
                },
              ],
              "type": "p",
            },
            Object {
              "instance": null,
              "key": null,
              "nodeType": "host",
              "props": Object {
                "children": "Lorem ipsum dolor sit amet.",
                "className": "description text-basic",
              },
              "ref": null,
              "rendered": "Lorem ipsum dolor sit amet.",
              "type": "p",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "host",
      "props": Object {
        "children": <div
          id="film-details"
        >
          <div
            className="poster"
          >
            <img
              alt="poster"
              src="https://image.tmdb.org/t/p/w640/img1234_0987"
            />
          </div>
          <div
            className="info"
          >
            <h2>
              <span
                className="title text-primary"
              >
                Pulp Fiction
              </span>
              <span
                className="rating"
              >
                8.3
              </span>
            </h2>
            <p
              className="genre text-basic"
            >
              Thriller
              Crime
            </p>
            <p
              className="numbers"
            >
              <span
                className="release-year"
              >
                1999
              </span>
              <span
                className="duration"
              >
                154
              </span>
            </p>
            <p
              className="description text-basic"
            >
              Lorem ipsum dolor sit amet.
            </p>
          </div>
        </div>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="poster"
            >
              <img
                alt="poster"
                src="https://image.tmdb.org/t/p/w640/img1234_0987"
              />
            </div>,
            <div
              className="info"
            >
              <h2>
                <span
                  className="title text-primary"
                >
                  Pulp Fiction
                </span>
                <span
                  className="rating"
                >
                  8.3
                </span>
              </h2>
              <p
                className="genre text-basic"
              >
                Thriller
                Crime
              </p>
              <p
                className="numbers"
              >
                <span
                  className="release-year"
                >
                  1999
                </span>
                <span
                  className="duration"
                >
                  154
                </span>
              </p>
              <p
                className="description text-basic"
              >
                Lorem ipsum dolor sit amet.
              </p>
            </div>,
          ],
          "id": "film-details",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": null,
            "nodeType": "host",
            "props": Object {
              "children": <img
                alt="poster"
                src="https://image.tmdb.org/t/p/w640/img1234_0987"
              />,
              "className": "poster",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": null,
              "nodeType": "host",
              "props": Object {
                "alt": "poster",
                "src": "https://image.tmdb.org/t/p/w640/img1234_0987",
              },
              "ref": null,
              "rendered": null,
              "type": "img",
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": null,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <h2>
                  <span
                    className="title text-primary"
                  >
                    Pulp Fiction
                  </span>
                  <span
                    className="rating"
                  >
                    8.3
                  </span>
                </h2>,
                <p
                  className="genre text-basic"
                >
                  Thriller
                  Crime
                </p>,
                <p
                  className="numbers"
                >
                  <span
                    className="release-year"
                  >
                    1999
                  </span>
                  <span
                    className="duration"
                  >
                    154
                  </span>
                </p>,
                <p
                  className="description text-basic"
                >
                  Lorem ipsum dolor sit amet.
                </p>,
              ],
              "className": "info",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": null,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    <span
                      className="title text-primary"
                    >
                      Pulp Fiction
                    </span>,
                    <span
                      className="rating"
                    >
                      8.3
                    </span>,
                  ],
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": null,
                    "nodeType": "host",
                    "props": Object {
                      "children": "Pulp Fiction",
                      "className": "title text-primary",
                    },
                    "ref": null,
                    "rendered": "Pulp Fiction",
                    "type": "span",
                  },
                  Object {
                    "instance": null,
                    "key": null,
                    "nodeType": "host",
                    "props": Object {
                      "children": "8.3",
                      "className": "rating",
                    },
                    "ref": null,
                    "rendered": "8.3",
                    "type": "span",
                  },
                ],
                "type": "h2",
              },
              Object {
                "instance": null,
                "key": null,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    "Thriller",
                    "Crime",
                  ],
                  "className": "genre text-basic",
                },
                "ref": null,
                "rendered": Array [
                  "Thriller",
                  "Crime",
                ],
                "type": "p",
              },
              Object {
                "instance": null,
                "key": null,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    <span
                      className="release-year"
                    >
                      1999
                    </span>,
                    <span
                      className="duration"
                    >
                      154
                    </span>,
                  ],
                  "className": "numbers",
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": null,
                    "nodeType": "host",
                    "props": Object {
                      "children": "1999",
                      "className": "release-year",
                    },
                    "ref": null,
                    "rendered": "1999",
                    "type": "span",
                  },
                  Object {
                    "instance": null,
                    "key": null,
                    "nodeType": "host",
                    "props": Object {
                      "children": "154",
                      "className": "duration",
                    },
                    "ref": null,
                    "rendered": "154",
                    "type": "span",
                  },
                ],
                "type": "p",
              },
              Object {
                "instance": null,
                "key": null,
                "nodeType": "host",
                "props": Object {
                  "children": "Lorem ipsum dolor sit amet.",
                  "className": "description text-basic",
                },
                "ref": null,
                "rendered": "Lorem ipsum dolor sit amet.",
                "type": "p",
              },
            ],
            "type": "div",
          },
        ],
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
